vulnfeeds/cmd/cvelist2osv/main.go

```diff
@@ -59,26 +65,419 @@ func extractConversionMetrics(cve cves.CVE5, refs []osvschema.Reference) {
+			nA.DatabaseSpecific["CPEs"] = vulns.Unique(cpeStrings)
+			v.Affected = append(v.Affected, nA)
+		}
+	}
+
+	// Combine 'affected' entries from both CNA and ADP containers.
+	cna := cve.Containers.CNA
+	adps := cve.Containers.ADP
+	gotVersions := false
+
+	affected := cna.Affected
+	for _, adp := range adps {
+		if adp.Affected != nil {
+			affected = append(affected, adp.Affected...)
+		}
+	}
+
+	// Attempt to extract version ranges from the combined 'affected' fields.
+	for _, cveAff := range affected {
+		versionRanges, versionType, extractNotes := ExtractVersionsFromAffectedField(cveAff, cve.Metadata.AssignerShortName)
+		notes = append(notes, extractNotes...)
+		if len(versionRanges) == 0 {
+			continue
+		}
+		gotVersions = true
+
+		if versionType == "git" {
+			nA := osvschema.Affected{}
+			for _, vr := range versionRanges {
+				vr.Type = osvschema.RangeGit
+				vr.Repo = cveAff.Repo
+				nA.Ranges = append(nA.Ranges, vr)
+			}
+			v.Affected = append(v.Affected, nA)
+		} else {
+			nA := osvschema.Affected{}
+			for _, vr := range versionRanges {
+				vr.Type = osvschema.RangeEcosystem
+				nA.Ranges = append(nA.Ranges, vr)
+			}
+			// Special handling for Linux kernel CVEs.
+			if cve.Metadata.AssignerShortName == "Linux" {
+				nA.Package = osvschema.Package{
+					Ecosystem: string(osvschema.EcosystemLinux),
+					Name:      "Kernel"}
+			}
+			v.Affected = append(v.Affected, nA)
+		}
+	}
+
+	// If no versions were found in 'affected', fall back to CPEs.
+	if !gotVersions {
+		notes = append(notes, "No versions in affected, attempting to extract from CPE")
+		cpeRanges, cpeStrings, err := findCPEVersionRanges(cve)
+		if err != nil {
+			notes = append(notes, err.Error())
+		}
+		if len(cpeRanges) != 0 {
+			nA := osvschema.Affected{}
+			for _, vr := range cpeRanges {
+				vr.Type = osvschema.RangeEcosystem
+				nA.Ranges = append(nA.Ranges, vr)
+			}
+			nA.DatabaseSpecific = make(map[string]interface{})
+			nA.DatabaseSpecific["CPEs"] = vulns.Unique(cpeStrings)
+			v.Affected = append(v.Affected, nA)
+		}
+	}
+
+	// As a last resort, try extracting versions from the description text.
+	if !gotVersions {
+		notes = append(notes, "No versions in CPEs so attempting extraction from description")
+		versions, extractNotes := cves.ExtractVersionsFromText(nil, cves.EnglishDescription(cve.Containers.CNA.Descriptions))
+		notes = append(notes, extractNotes...)
+		if len(versions) > 0 {
+			// NOTE: These versions are not currently saved due to the need for better validation.
+			notes = append(notes, fmt.Sprintf("Extracted versions from description but did not save them: %+v", versions))
+		}
+	}
+
+	return notes
+}
+
+// addToVersionRange is a helper function that adds 'introduced', 'fixed', or 'last_affected'
+// events to an OSV version range. If 'intro' is empty, it defaults to "0".
+func addToVersionRange(versionRange *osvschema.Range, intro string, lastAff string, fixed string) {
```

Comment:
Hmm, looking at all the usages of addToVersionRange, you seem to be using it with a brand new versionRange every time. Is it possible to just have this return a new versionRange, and rename it buildVersionRange or something. 

I'm guessing this would never add additional events to an existing range, because you always add an introduce every time this function is called, and it's not valid to have multiple introduce in one version range.